### ISSUE TITLE
Can the UI be forced to use relative references for weather file location?

### ISSUE THREAD

**Issue by sno036** [APSIMInitiative/ApsimX#7120]
Would be useful!!

**Comment by hol430** [comment]
This should already be happening. The only requirement is that the .met file needs to be in the same directory tree (ie in the same directory or a subdirectory) of the .apsimx file.

I suppose we could start inserting ".." characters to force always using a relative directory, but this would break if the user ever moved the .apsimx file.

**Comment by lie112** [comment]
I believe the open file dialog of the property presenter provides the full path of the file selected to the property. Should it trim to a path relative to apsimx file if paths match? Often when I get a user's simulation file I have to strip all their local drive text from file names added by file property dialog.

**Comment by hol430** [comment]
The open file dialog doesn't know anything about the .apsimx file (nor should it, I think), so it has no way of generating a relative path. Its only purpose is to ask a user for a file name and return that file name.

It doesn't make sense to use a relative path in all cases anyway - e.g. when opening an .apsimx file what would the path be relative to? I'd rather keep the file chooser as simple as possible and have it do one task well.

That said, it should be easy enough to write your own extension method for the file dialog which returns a relative path. E.g.

```csharp
public static string GetRelativeFile(this FileDialog dialog, string relativePath)
{
    string path = dialog.GetFile();
    return PathUtilities.GetRelativePath(path, relativePath);
}
```

**Comment by lie112** [comment]
Fair enough. Does make passing files and zipped working simulations+data between people a little tricky. You either have to remember to provide a relative path, or fix the file not found error when thrown.

**Comment by sno036** [comment]
@hol430 - that might be a purest developer point of view perhaps? From my (user!) point of view the open file dialogue should know the location of the .apsimx file as that is where the file needed is mostly likely to be located. Hard coded file paths are painful when sharing simulations and expensive when sending jobs to the cloud and it needs fixing.

**Comment by peter-devoil** [comment]
If we're thinking of sharing/exchange - why can't we encode a met file into json?

**Comment by hol353** [comment]
The current situation is difficult for the user and often requires manually editing the .apsimx file.

**Use case**: The SWIM simulations we are working on reference the lincoln weather file in examples. Before we raise a PR we have to remember to go and change the weather paths (in a text editor) because they are absolute paths and won't run on Jenkins. 

**Three solutions**:
1. Write a %apsimng% style of macro if the weather file is located under the apsimx directory structure.
2. Use relative paths if possible i.e. if the .met file is in the same directory structure as the .apsimx file.
3. Encode the contents of the weather file into the .apsimx file.

Perhaps we need a combination of these. i.e. the GUI tries option 1 first. If not possible, go for option 2. If not possible, write absolute paths. The GUI could also have a button in the GUI to do option 3.

**Comment by her123** [comment]
Another option is to have either the application, simulation object or the weather object try to locate the file specified (at opening time or at start of run). If it is not found at the path then try to find it in the same path as the .apsimx file. Or possibly going further and trying sub-directories.

**Comment by hol430** [comment]
> 1. Write a %apsimng% style of macro if the weather file is located under the apsimx directory structure.

We already have a `%root%` macro which will be resolved at runtime to the apsimx directory (parent of the directory in which the ApsimNG executable is located). But, consider the following scenario:

Let's say I'm running the released version, but I have also cloned apsim on my computer, so that I can access the prototypes. Let's say I use the released version and open up the weirdo prototype. The met file will be pointing to %root%\Prototypes\WEIRDO\lincoln.met. %root% will be resolved to "C:\Program Files\Apsimxxxx.xx.xx.xx". However, the Prototypes directory is not shipped with the installed version, so apsim will not be able to find the met file at %root%\Prototypes\WEIRDO\lincoln.met.

In order to accommodate this use case, the `%root%` macro will never be inserted automatically by the gui and needs to be done manually if you want to use it.

> 2. Use relative paths if possible i.e. if the .met file is in the same directory structure as the .apsimx file.

This is what we already do for .met files (with the proviso from my earlier comment), except that it's handled internally by the weather component rather than the GUI. I would like to see this generalised for all input files but I'm not sure of the best way to do this, and I definitely don't think it makes sense to do this at the file dialog level. One problem of using relative paths is that the model itself needs some smarts in order to resolve the relative file path from the path of the .apsimx file. We could perhaps create a class to handle this and use that instead of the string data type for these relative paths. Another problem is that it's not always possible to create a relative path between two files (think files stored on different drives in windows).

> 3. Encode the contents of the weather file into the .apsimx file.

Do you mean change the .apsimx file format to (for example) a tarball with the .apsimx extension? I actually don't mind that idea, you could embed a bunch of files in there if need be. Main downside would be that you're wasting disk space by having the same file contents stored in multiple places. Of course we could always compress the .apsimx file before writing it.

**Comment by hol353** [comment]
> We already have a %root% macro which will be resolved at runtime to the apsimx directory (parent of the directory in which the ApsimNG executable is located).

I think this is the problem. We need to make %root% relative to the apsim root of the directory where the .apsimx file is located rather than the root of the apsim directory where the gui is running i.e. %root% is resolved to wherever the parent apsimx directory that contains the prototypes directory (not C:\Program Files...). Will be a bit tricky to figure out - perhaps use the presence of a 'special file' e.g. APSIM.bib. Does this solve the problem?

2. In my SWIM use case, we were working out of a cloned apsim directory. The .apsim file was in Apsimx\Prototypes\Swim.apsim. The weather file should be a relative link: ..\..\Examples\WeatherFiles\lincoln.met. Instead it was written as an absolute path. Bug? I like the idea of a class that handles this mess and use an instance of it rather than a string.
3. Yes, we could support storing the contents into the JSON just as string lines:
```
        {
          "$type": "Models.Weather, Models",
          "Name": "Weather",
          "Data": [
                      "  year    day     radn    maxt    mint    rain",
                      "()      ()      (mJ/m^2)  (oC)    (oC)    (mm)",
                      "  1995     1     16.85    31.2    20.2    0.00",
                      "  1995     2     16.85    31.4    19.8    0.00",
                      "  1995     3     16.85    30.2    19.6    5.30"
           ]
        },
```
This is crude but it can be done. It will result in larger .apsimx files but we can always zip compress them. Whatever we do, we still need to support users manually editing them.

**Comment by sno036** [comment]
#3 encode the met data - think of an experiment node referencing 50 locations - hell make it national grid of 50,000.
#2 yes good that all operate the same way
#1 hard nup from me with respect to the searching multiple locations. It needs to be unambiguously clear what met file was used. But yes about the %root% as Dean writes.

**Comment by hol430** [comment]
> We need to make %root% relative to the apsim root of the directory where the .apsimx file is located rather than the root of the apsim directory where the gui is running

The whole point of the `%root%` macro (it seems to me) is so that the user can open an example file, save it somewhere else on their computer, and run it. If we change the `%root%` as you describe, it is only resolve-able if the .apsimx file is saved somewhere in an apsimx repo.

Even if we were happy to remove this feature, detecting an apsimx "repo"  as you describe it is not so easy, as the directory can look different when cloned vs installed (the windows, mac and linux installations all look a little different). I also don't like relying on the existence of a file - it's easy to imagine users tripping themselves up by having a file somewhere with the same name we're looking for. A feature like this would also make it more difficult to change the directory structure of the repository in the future.

> The weather file should be a relative link: ....\Examples\WeatherFiles\lincoln.met. Instead it was written as an absolute path. Bug?

At the moment, this is "intentional" (see my first comment in this thread), but I'll agree it's perhaps less than ideal. We could change this to always use a relative path (when possible). The problem is that if you move the .apsimx file and forget to move the .met file, you won't be able to run your sim. I guess it's easier to forget to move the .met file as well if it's stored at `../../../../../../weatherfiles/file.met`.

> we could support storing the contents into the JSON just as string lines

I don't really like the idea of storing the raw lines as strings in the json. If we were to go down that path, I'd rather store parsed objects instead. What I was imagining though was more like storing multiple files (.json, .met, .xlsx, ...) inside a single compressed archive which has the .apsimx file extension. People would still be able to edit these if they need to by cracking them open with their archiving tool of choice. I'd imagine .met files would compress down pretty well like this, and I'd be surprised if 50k met files took more than a couple hundred MB compressed. This sounds like a lot of work for us to implement though so it's probably a moot point.

**Comment by jbrider** [comment]
possible solution: move to .apsimz file format that is a zipped package of supporting files.
If you need to link to an external file ie: many apsim files referencing a common met file, then be able to handle a file reference as the exception not the norm.
I believe this is how microsoft now deals with it's file formats - powerpoint files definitely work like this.
I wouldn't store the db files in the package by default - but it would be handy to have the option when exporting them.

**Comment by zur003** [comment]
Some comments from a retired old fart. I remember trying to address this issue a decade or so ago, and eventually giving up because I couldn't find a solution that handled all use-cases well. I think the main issues have already been described above.

Embedding weather data in the .apsimx file could only ever be an optional way of doing things. Series of simulations running at multiple locations (like the World Modellers stuff) really requires being able to access weather data from an external (likely freshly downloaded) source. Similarly for simulations wanting to use the very latest data from SILO or wherever. However, some sort of embedding option might be useful for dealing with example simulations that are likely to be copied by the user to a new location.
